implement seed mechanism

domcyrus opened the pull request.

shreddedbacon reviewed (changes requested):
This is all good. Only concern I have is that a `DatabaseRequest` that has fails to seed continues to be reconciled flooding the logs with reconcile errors.

This seems unnecessary. The `DatabaseRequest` should also probably go into a `failed` status that is easier to see
```
status:
  conditions:
    - lastTransitionTime: '2024-06-05T03:46:20Z'
      message: >-
        mysql db find connection from seed failed
      reason: Failed
      status: 'True'
```

domcyrus commented on internal/database/database.go:
```diff
@@ -112,6 +117,32 @@ func (ri *RelationalDatabaseImpl) Ping(ctx context.Context, dsn string, dbType s
 
 	if err := db.PingContext(ctx); err != nil {
+		if dbType == mysql {
+			log.FromContext(ctx).Error(err, "Failed to ping MMMMMySQL database")
```
Will remove... It was some leftover from some testing.

domcyrus commented:
> This is all good. Only concern I have is that a `DatabaseRequest` that has fails to seed continues to be reconciled flooding the logs with reconcile errors.
> 

Regarding flooding the logs:

The `DatabaseRequestReconciler` will get retried using the normal controller-runtime ratelimit. If we want to tweak that we could set the ratelimiter `RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(baseDelay, maxDelay)`. 

Another idea would be to not log anything if nothing in the seed has changed and the error in the status condition is related to the seed. In that case the `DatabaseRequestReconciler` needs to store the seed and compare it which is possible but maybe a bit brittle.

I would vote to rather keep the current default runtime behavior which will solve handle the same error by increasing the time to retry. It will log a few times the same error at the beginning but later on it won't do that often anymore. I think this is a good behavior because it fixes issues automatically and plays nice with kubernetes eventual consistency concept.

> This seems unnecessary. The `DatabaseRequest` should also probably go into a `failed` status that is easier to see
> 
> ```
> status:
>   conditions:
>     - lastTransitionTime: '2024-06-05T03:46:20Z'
>       message: >-
>         mysql db find connection from seed failed
>       reason: Failed
>       status: 'True'
> ```
I've added an additional Error status condition.

shreddedbacon commented:
> > This is all good. Only concern I have is that a `DatabaseRequest` that has fails to seed continues to be reconciled flooding the logs with reconcile errors.
> 
> Regarding flooding the logs:
> 
> The `DatabaseRequestReconciler` will get retried using the normal controller-runtime ratelimit. If we want to tweak that we could set the ratelimiter `RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(baseDelay, maxDelay)`.
> 
> Another idea would be to not log anything if nothing in the seed has changed and the error in the status condition is related to the seed. In that case the `DatabaseRequestReconciler` needs to store the seed and compare it which is possible but maybe a bit brittle.
> 
> I would vote to rather keep the current default runtime behavior which will solve handle the same error by increasing the time to retry. It will log a few times the same error at the beginning but later on it won't do that often anymore. I think this is a good behavior because it fixes issues automatically and plays nice with kubernetes eventual consistency concept.
> 
> > This seems unnecessary. The `DatabaseRequest` should also probably go into a `failed` status that is easier to see
> > ```
> > status:
> >   conditions:
> >     - lastTransitionTime: '2024-06-05T03:46:20Z'
> >       message: >-
> >         mysql db find connection from seed failed
> >       reason: Failed
> >       status: 'True'
> > ```
> 
> I've added an additional Error status condition.

I think this behaviour and lagoon are a bad combination generally. But I think we can be a bit smart about the things that retry and continue trying to reconcile until success, and things that are deemed a hard failure and to not retry them with.

Some parts of a Lagoon build are more suited to a hard failure than a "wait for" condition, that's all.

domcyrus commented:
I understand. If we prefer to return success and not retry on an invalid or broken seed, that works fine for me.

I was thinking that it might make sense to retry. For example, if a platform engineer is alerted, they could inspect the seed secret and fix it manually. In that case retrying would allow the system to automatically correct itself. Without retrying, the system will not have the opportunity to autocorrect.

shreddedbacon commented:
> I was thinking that it might make sense to retry. For example, if a platform engineer is alerted, they could inspect the seed secret and fix it manually. In that case retrying would allow the system to automatically correct itself. Without retrying, the system will not have the opportunity to autocorrect.

In theory this is fine. But it could end very badly if something is done incorrectly. In this case, if a seed fails it should not retry. The lagoon build will drop a warning and continue to use the `MariaDBConsumer`. The warning will inform of a failure to convert and require the customer to ask for help.

If it was to continue to retry, and someone from platform stepped in and resolved the issue incorrectly, then it could potentially result in outages or downtimes. A failure to convert from `MariaDBConsumer` to `DatabaseRequest` is not something that would be resolvable without involving a customer basically.

Because of some very old bugs in Lagoon builds, there are some `MariaDBConsumers` out there that are incorrectly provisioned.

I'll have to put together documentation about how to recover from these situations though when we get closer to doing this within a build though, because it could impact opensource users as well as AIO directly.

domcyrus commented:
I've changed the code to not retry on specific seed errors. This triggered tests on seed to randomly fail, because of a race condition when creating the seed database. Anyway I've fixed it with a sleep. I think it illustrates a good reason why retries are useful.

shreddedbacon commented:
> I've changed the code to not retry on specific seed errors. This triggered tests on seed to randomly fail, because of a race condition when creating the seed database. I've fixed it with a sleep. I think it makes a good example of why retries are useful.
> 
> PS: I'll certainly remove the comments about the retry it is just to illustrate on why they might be useful.

Since the sleep is only in the test suite to give the mysql pod time to start up for the test to run, then this is somewhat acceptable to me, I'm sure the test could also be refactored to ensure the pod is running though instead of an arbitrary sleep too.

domcyrus commented:
@shreddedbacon Is there anything else needed?

shreddedbacon reviewed (approved):
Did some quick verifications with emulating how a build may create the seed and request and this works great. The failure case with the seed is ideal for how Lagoon builds work to prevent unexpected behaviours.

Thanks Marco!
